streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/StreamLoader.scala

```diff
@@ -0,0 +1,144 @@
+  //                                                   +-------------| sink     |
+  //                                                                 +-----|----+
+  //                                                                       |
+  //                                                                       |                 +--------+
+  //                                                                       |     insert      |BigQuery|
+  //                                                                       +-----------------|(events)|
+  //                                                                                         +--------+
+
+  private val MaxConcurrency = Runtime.getRuntime.availableProcessors * 16
+  private val GroupByN       = 10
+  private val TimeWindow     = 5.minutes
+
+  implicit val cs: ContextShift[IO] = IO.contextShift(scala.concurrent.ExecutionContext.Implicits.global)
+
+  def run(e: Environment)(implicit cs: ContextShift[IO], c: Concurrent[IO], T: Timer[IO]): IO[ExitCode] = {
+    val source = new PubsubSource(e)(cs, c)
+
+    val eventStream: Stream[IO, Either[BadRow, StreamLoaderRow]] =
+      source.getStream.evalMap(StreamLoaderRow.parse(e.resolverJson))
+
+    for {
+      client <- Bigquery.getClient
+      queue  <- Queue.bounded[IO, Set[ShreddedType]](MaxConcurrency)
+      sinkBadGood = eventStream.observeEither[BadRow, StreamLoaderRow](
+        badSink(e),
+        goodSink(queue, _.inventory, bigquerySink(e, client))
+      )
+      sinkTypes = queue.dequeue.through(aggregateTypes(GroupByN, TimeWindow)).through(typesSink(e))
+      _ <- Stream(sinkTypes, sinkBadGood).parJoin(MaxConcurrency).compile.drain
+    } yield ExitCode.Success
+  }
+
+  // TODO: Maybe sinks can be generalised?
+
+  /**
+    * Sink bad rows to Pub/Sub.
+    * @param e Parsed common environment.
+    * @return A sink that writes bad rows to Pub/Sub.
+    */
+  def badSink(e: Environment): Pipe[IO, BadRow, Unit] =
+    _.parEvalMapUnordered(MaxConcurrency)(badRow =>
+      PubSub.write(PubSubOutput.WriteBadRow(badRow), e.config.projectId, e.config.badRows)
+    )
+
+  /**
+    * Sink types to Pub/Sub.
+    * @param e Parsed common environment.
+    * @return A sink that writes observed types to Pub/Sub (to be consumed by Mutator).
+    */
+  def typesSink(e: Environment): Pipe[IO, Set[ShreddedType], Unit] =
+    _.parEvalMapUnordered(MaxConcurrency)(typesSet =>
+      PubSub.write(PubSubOutput.WriteObservedTypes(typesSet), e.config.projectId, e.config.typesTopic)
+    )
+
+  /**
+    * Sink rows to BigQuery and failed inserts to Pub/Sub.
+    * @param e Parsed common environment.
+    * @param c BigQuery client.
+    * @return A sink that writes rows to a BigQuery table and routes failed inserts to Pub/Sub (to be consumed by Repeater).
+    */
+  def bigquerySink(e: Environment, c: BigQuery): Pipe[IO, StreamLoaderRow, Unit] =
+    _.parEvalMapUnordered(MaxConcurrency)(loaderRow => Bigquery.insert(loaderRow, e, c))
```

Comment:
What happens if the write to BigQuery fails ? Is there a retry mechanism ? Is the message ACKed ?